nerpa_controller/src/lib.rs

```diff
@@ -1069,13 +1098,23 @@ impl ControllerActor {
                     }
 
                     let ddlog_res = self.program.apply_updates(vec![inp_opt.unwrap()]);
-                    if ddlog_res.is_ok() {
-                        let p4_res = self.switch_client.push_outputs(&ddlog_res.unwrap()).await;
+                    if ddlog_res.is_err() {
+                        error!("could not apply changes to ddlog input relation");
+                    }
+
+                    let ddlog_output = &ddlog_res.unwrap();
+
+                    // We use the first switch client to construct the updates and packets
+                    // that we then send to the switch.
+                    // Assumes that all switches are running the same P4 program.
+                    let (updates, packets) = self.switch_clients[0]
```

Comment:
This can't be right, switches cannot share the same configuration.
The ddlog output tables should contain a column identifying the client.

Reply:
Perhaps this part is fine if you assume that all switches run the same P4 code, but the loop below cannot be right, updates should be per switch.

Reply:
Got it, this makes sense. It is easy to revert to per-switch updates.

If DDlog output tables contain a column with the client, do all input relations and control-plane rules also require a client column? This seems to increase the Nerpa programmer's work. Is there any easier way to do this?

Reply:
I don't think all input relations need a client column, because the user doesn't necessarily configure the network on a per-switch basis.

I think that input relations that relate to a particular switch do need to designate the switch. For example, we'd need it for packets that were received from a switch.

I do think that there will be tables that could be the same on every switch in some applications so that they don't necessarily need a column for the switch. OVN, for example, installs essentially the same forwarding table on every hypervisor. I guess that's an optimization we can implement later.

Reply:
This is quite amazing that all hypervisors have the same forwarding rules. Where are then the routing tables that take into account the network topology? Only in "real" switches?

Reply:
They are not completely identical. The forwarding pipeline on each hypervisor looks sort of like this:
1. A few forwarding tables that transform physical packets into a kind of logical forwarding space.
2. Most of the forwarding tables, which make decisions for the packet on the basis of its logical properties
3. A few forwarding tables that transform the logical properties back into physical ones.

The tables in 1 and 3 differ from one HV to another, but the ones in 2 are the same, although some of them can be dropped, as an optimization, on HVs that won't ever see packets that would match them.

Reply:
Based on the above, it seems that the output relations generated by `p4info2ddlog` and the input relations generated from OVSDB do not need a client column. They would be the same across all switches.

But input relations generated from the data plane (corresponding to packets, say) do require a client column. The nerpa controller should only send packets to that client.

Is this correct?